File: 2022/june/29-06-2022/queue-reconstruction-by-height.py
class Solution:
    def reconstructQueue(self, people):
        if len(people) <= 1:
            return people

        people = sorted(people, key=lambda x: (-x[0], x[1]))

        for cur in range(len(people)):
            h, k = people[cur]

            if k < cur:
                p = people.pop(cur)
                people.insert(k, p)

        return people


print(Solution().reconstructQueue([[7,0],[4,4],[7,1],[5,0],[6,1],[5,2]]))

File: 2022/june/29-06-2022/queue-reconstruction-by-height.rs
const Height: usize = 0;
const Rank: usize = 1;

impl Solution {
    pub fn reconstruct_queue(mut people: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        // Sort people by decreasing height, and increasing rank:
        people.sort_by_key(|p| (-p[Height], p[Rank]));
        // Reconstruct the queue:
        let mut queue = Vec::with_capacity(people.len());
        for person in people.iter() {
            queue.insert(person[Rank] as usize, person.to_vec());
        }
        queue
    }
}
